devenv-run-tests/src/main.rs

```diff
@@ -71,30 +71,55 @@ fn run_tests_in_directory(args: &Args) -> Result<Vec<TestResult>, Box<dyn std::e
                     }
                 }
 
-                println!("  Running {}", dir_name);
-                // if .setup.sh exists, run it
-                let setup_script = path.join(".setup.sh");
-                if setup_script.exists() {
-                    println!("    Running .setup.sh");
-                    let _ = std::process::Command::new("bash")
-                        .arg(".setup.sh")
-                        .current_dir(&path)
-                        .status()?;
-                }
                 let overrides = args.override_input.iter().enumerate().flat_map(|(i, arg)| {
                     if i % 2 == 0 {
                         vec!["--override-input", arg.as_str()]
                     } else {
                         vec![arg.as_str()]
                     }
                 });
+
+                let devenv_override = [
+                    "--override-input",
+                    "devenv",
+                    &format!("path:{cwd}?dir=src/modules"),
+                ];
+
+                // A script to patch files in the working directory before the shell.
+                let patch_script = ".patch.sh";
+                let patch_script_path = path.join(patch_script);
+
+                // A script to run inside the shell before the test.
+                let setup_script = ".setup.sh";
+                let setup_script_path = path.join(setup_script);
+
+                println!("  Running {}", dir_name);
+
+                // Run .patch.sh if it exists
+                if patch_script_path.exists() {
+                    println!("    Running {patch_script}");
+                    let _ = std::process::Command::new("bash")
+                        .arg("./.patch.sh")
+                        .current_dir(&path)
+                        .status()?;
+                }
+
+                // Run .setup.sh if it exists
+                if setup_script_path.exists() {
+                    println!("    Running {setup_script}");
+                    // TODO: use as a library
+                    let _ = std::process::Command::new("devenv")
+                        .args(devenv_override)
+                        .args(overrides.clone())
+                        .arg("shell")
+                        .arg("./.setup.sh")
```

Comment:
So this is the same as `enterShell`, but it's not racy so that's why it works better?

Reply:
Exactly. The problem we have now is that this sets things up in `.devenv`, but then `devenv test` uses a temporary directory. We can work around this by librarifying devenv and setting up our own temp dir.